### ISSUE TITLE
Question: does katago do an position merging?

### ISSUE THREAD

**Issue by bhansconnect** [lightvector/KataGo#472]
The title says most of it. I am just wondering if katago does any sort of position merging. I saw that for connect4 at least it was very helpful due to the opening being repeated so often. I would assume for go, the possible first 4ish moves are extremely repetitive.

Also, if you don't use it, why not/what are your thoughts on it?

I am working on my own ai, and was wondering the merrit of it. I guess my main concern about it is that avoiding merging means that you have many more opening samples, which arguable would be good to over sample because it is important to play them well. With merging you don't value any position more than others.

**Comment by lightvector** [comment]
Hmmm, what is position merging? Is it some method of sampling opening moves that reduces the frequency of sampling early moves more? Or are you referring to taking multiple SGF game files and merging them into a tree or something so they share their first few moves in the SGF file? Or do you mean you prohibit ever observing the same position twice - once you've seen on a given position in self-play and decided on a policy, when you see it again the next game you skip it and replay the policy, and you never try that position again until you produce the next generation of neural net?

In self-play, KataGo is currently configured to take a draw from a geometric distribution roughly a mean of 4% of the area of the board, and plays that many moves using the raw temperature-1 policy to initialize the board, before starting the game for real, to provide some opening diversity. The opening diversity is the more important focus, but as a side effect it means the first few moves of a game are much more under-trained on relative to their frequency (but still over-trained relative to other positions), I think addresses your concern to some degree too.

**Comment by bhansconnect** [comment]
So this of course will be worse for connnect4 than go due to the limited action space, but essentially they took their dataset from selfplay and averaged the policy and value results for each position. So for example, in training, the ai might have played the 4 4 point 10,000 times. By default, that would be 10,000 different training samples. What oracle did was average the policy and value together for all 10,000 samples and made it into a single training sample.

The description in [here](https://medium.com/oracledevs/lessons-from-alpha-zero-part-6-hyperparameter-tuning-b1cfcbe4ca9a) under position averaging.

**Comment by lightvector** [comment]
Ah. Yeah, my intuition is that technique is probably worth less outside of small toy-sized games. And connect 4 is definitely a toy-sized game. In larger games a higher proportion of your data is going to be unique, which means you have less and less effect. 

Also, I wonder if the technique fixes the problem the wrong way. If you have a huge amount of repeated opening positions in your training data instead of spending the costly compute time to generate all the data for those positions over and over again, it seems like you could just pick any number of methods to avoid wasting the compute to generate so much of that repeated data in the first place. Repeating it somewhat will of course help the neural net average over the noise, but you're way past overkill when 30% of your data is on a limited set of duplicated opening positions.

For example, sample positions from an automatically-generated opening book so as to frequently skip past those positions that you'd be repeating. Or use something like KataGo's exponentially weighted fast raw-policy opening initialization that I described above, which is kinda like using the raw policy as a probabilistic book. Or still play the opening as normal but with extra-high move choice temperature for the opening so as to get more variety and rapidly branch into unique positions. Or use game forking like LC0 or SAI have tried where exploration branches the game state (i.e. if on move 12 you want to play an exploration move, you branch into 2 games, one where you play the top move, one where you play the exploration move) so that because trajectories branch and grow bushy as you go outward, the early opening is a smaller proportion of your data.

**Comment by bhansconnect** [comment]
Thanks for the explanation. It makes a lot of sense. A follow up question, if you know, how much do exponentially weighted fast raw-policy opening initialization and game forking end up effecting training and final elo? Do they have a significant impact in the long term?

**Comment by mega-optimus** [comment]
reducing repeated samples to 1 changes the sample distribution, this is definitely harmful.

**Comment by lightvector** [comment]
> Thanks for the explanation. It makes a lot of sense. A follow up question, if you know, how much do exponentially weighted fast raw-policy opening initialization and game forking end up effecting training and final elo? Do they have a significant impact in the long term?

I don't have stats, but to get reliable convergence to optimal for small-medium-sized games (e.g. 7x7 Go) and to get good training on larger games, I'd guess you generally need at least one solid form of exploration encouragement, and ideally a few. I'm not sure I've seen anyone specifically try to ablate away all their core exploration and game-diversity methods, which most of these methods are, but you might be able to search online and find some results in the many papers out there replicating AlphaZero-like experiments.

It's common to also add a softmax temperature at the root > 1 so that the MCTS at the root also explores a bit wider than the policy probabilities, and results in an equilibrium policy that doesn't converge oversharply. 

At temperature 1, if you see a position repeatedly over the course of successive nets, and iterate the cycle of policy -> training target from mcts -> new policy fitted on it -> new training target from mcts -> ..., almost any deviation in winrate, no matter how slight, in theory will cause the equilibrium policy to be 100% on the top move and epsilon% on all other moves, where epsilon% is a tiny value due to dirichlet noise playouts. For non-toy-sized action spaces, this mean the policy converges way too rapidly in practice, and the winrates early in training driving this premature convergence are not necessarily correct if the net hasn't learned enough yet. A temperature > 1 results in an equilibrium where there will be more than epsilon probability on any moves that are "close" in winrate. You can simulate this in theory if you like - just write a python script that does 1-ply MCTS on a position whose winrates of each action are fixed to some constant, a bunch of times, generates a playout distribution, and then plugs the average distribution back in as the policy prior (simulating a neural net that converges perfectly on it), and repeats.

**Comment by bhansconnect** [comment]
Thanks for all of the information. It is very useful.
